### PR TITLE
[Evaluator][Feature Branch] Implementation of `evaluate` function

### PR DESCRIPTION
**Pull request by dbogunowicz**
## Feature Description

This is a feature branch for the DeepSparse Evaluation API. This PR briefly introduces the main `evaluate` function.

The following work items are the DoD of this PR:

- [x] - Land #1388
- [x] - Land #1387
- [x] - Land #1390
- [x] - Land #1393
- [x] - Make sure that all the components implemented work harmoniously when put together
- [x] - Discuss the implemented `Evaluation API` with the product and introduce appropriate changes
- [x] - As a result of the two items mentioned above, land #1410 
- [x]  -...and #1411 
- [x] - Finalise the documentation

## Architecture Diagram
image

### CODE REVIEW

File: src/deepsparse/evaluation/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    dataset: Dataset
+    metrics: List[Metric]
+    samples: List[EvalSample]
+
+
+def evaluate(
+    target: str,
+    datasets: Union[str, List[str]],
+    integration: str,
+    evaluation_registry: BaseEvaluationRegistry,
```
Comment (bfineran) [inline src/deepsparse/evaluation/evaluator.py:60]: not needed

File: src/deepsparse/evaluation/registry.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Callable, List, Union
+
+
+class BaseEvaluationRegistry(ABC):
```
Comment (bfineran) [inline src/deepsparse/evaluation/registry.py:19]: move to RegistryMixin if possible

File: src/deepsparse/evaluation/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+                        of a torch model), the engine type should be set to None"
+    -b BATCH_SIZE, --batch_size BATCH_SIZE
+                        The batch size to use for the evaluation. Must be greater than 0
+    -s SPLITS, --splits SPLITS
+                        The name of the splits to evaluate on. Can be a string for a single split
+                        or a list of strings for multiple splits.
+    -m METRICS, --metrics METRICS
+                        The name of the metrics to evaluate on. Can be a string for a single metric
+                        or a list of strings for multiple metrics.
+    --enforce_result_structure --enforce-result-structure
```
Comment (dbogunowicz) [inline src/deepsparse/evaluation/evaluator.py:55]: include the original structure or not (we are always returning our structure)

File: src/deepsparse/evaluation/evaluator.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    )
+
+    eval_integration = EvaluationRegistry.load_from_registry(integration)
+
+    _LOGGER.info(
+        f"The following evaluation integration has "
+        f"been successfully setup: {eval_integration.__name__}"
+    )
+
+    result = eval_integration(
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/evaluation/evaluator.py:190]: add it to the Result object

File: src/deepsparse/evaluation/evaluator.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+            raise ValueError(
+                "The evaluation integration must return a list of Evaluation objects "
+                "when enforce_result_structure is True."
+            )
+        _LOGGER.info(f"Evaluation done. Results:\n{print_result(result)}")
+
+    return result
+
+
+def main():
```
Comment (dbogunowicz) [inline src/deepsparse/evaluation/evaluator.py:212]: API should not print, the CLI should. On the start and at the end. Option to save through the CLI

File: src/deepsparse/evaluation/utils.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    :param kwargs: Additional kwargs to pass to the model initialization
+    :return: The initialized model
+    """
+    if engine_type in [DEEPSPARSE_ENGINE, ORT_ENGINE]:
+        return Pipeline.create(
+            task="text-generation", model_path=target, engine_type=engine_type, **kwargs
+        )
+    elif engine_type == "torch":
+        return AutoModelForCausalLM.from_pretrained(target, **kwargs)
+    else:
```
Comment (dbogunowicz) [inline src/deepsparse/evaluation/utils.py:41]: user passes the model or pass torch script or sparseml.
Move to sparseml

File: src/deepsparse/evaluation/integrations/llm_evaluation_harness.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+def integration_eval(
+    target: str,
+    datasets: Union[List, str],
+    batch_size: int,
+    target_args: Dict[str, Any],
+    engine_type: Optional[str] = None,
+    **kwargs,
+) -> Dict[str, Any]:
+    """
+    Reimplementation of:
```
Comment (dbogunowicz) [inline src/deepsparse/evaluation/integrations/llm_evaluation_harness.py:58]: if engine_type = None assert native model framework

File: setup.py [hunk 8]
```diff
@@ -305,6 +305,7 @@ def _setup_entry_points() -> Dict:
             f"deepsparse.image_classification.eval={ic_eval}",
             "deepsparse.license=deepsparse.license:main",
             "deepsparse.validate_license=deepsparse.license:validate_license_cli",
+            "deepsparse.eval=deepsparse.evaluate.evaluator.evaluate:main",
```
Comment (dbogunowicz) [inline setup.py:308]:
```suggestion
            "deepsparse.eval=deepsparse.evaluate.evaluator.cli:main",
```

**Pull request merged**